tortoise/state.go

```diff
@@ -5,78 +5,242 @@ import (
-	weight util.Weight
+type (
+	weight = util.Weight
+
+	verifyingInfo struct {
+		good, abstained weight
+		referenceHeight uint64
+	}
+
+	layerInfo struct {
+		lid            types.LayerID
+		empty          weight
+		hareTerminated bool
+		blocks         []*blockInfo
+		verifying      verifyingInfo
+	}
+
+	blockInfo struct {
+		id       types.BlockID
+		layer    types.LayerID
+		height   uint64
+		hare     sign
+		validity sign
+		margin   weight
+	}
+
+	state struct {
+		// last received layer
+		// TODO should be last layer according to the clock
+		last types.LayerID
+		// last verified layer
+		verified types.LayerID
+		// historicallyVerified matters only for local opinion for verifying tortoise
+		// during rerun. for live tortoise it is identical to the verified layer.
+		historicallyVerified types.LayerID
+		// last processed layer
+		minprocessed, processed types.LayerID
+		// last evicted layer
+		evicted types.LayerID
+
+		localThreshold  weight
+		globalThreshold weight
+
+		// epochWeight average weight per layer of atx's that target keyed epoch
+		epochWeight map[types.EpochID]weight
+		// referenceHeight is a median height from all atxs that target keyed epoch
+		referenceHeight map[types.EpochID]uint64
+		// referenceWeight stores atx weight divided by the total number of eligibilities.
+		// it is computed together with refBallot weight. it is not equal to refBallot
+		// only if refBallot has more than 1 eligibility proof.
+		referenceWeight map[types.BallotID]weight
+
+		layers  map[types.LayerID]*layerInfo
+		ballots map[types.LayerID][]*ballotInfo
+
+		// to efficiently find base and reference ballots
+		ballotRefs map[types.BallotID]*ballotInfo
+		// to efficiently decode exceptions
+		blockRefs map[types.BlockID]*blockInfo
+	}
+)
+
+func newState() *state {
+	return &state{
+		epochWeight:     map[types.EpochID]util.Weight{},
+		referenceHeight: map[types.EpochID]uint64{},
+		referenceWeight: map[types.BallotID]util.Weight{},
+
+		layers:     map[types.LayerID]*layerInfo{},
+		ballots:    map[types.LayerID][]*ballotInfo{},
+		ballotRefs: map[types.BallotID]*ballotInfo{},
+		blockRefs:  map[types.BlockID]*blockInfo{},
+	}
+}
+
+func (s *state) layer(lid types.LayerID) *layerInfo {
+	layer, exist := s.layers[lid]
+	if !exist {
+		layer = &layerInfo{lid: lid, empty: util.WeightFromUint64(0)}
+		s.layers[lid] = layer
+	}
+	return layer
+}
+
+func (s *state) addBallot(ballot *ballotInfo) {
+	s.ballots[ballot.layer] = append(s.ballots[ballot.layer], ballot)
+	s.ballotRefs[ballot.id] = ballot
+}
+
+func (s *state) addBlock(block *blockInfo) {
+	layer := s.layer(block.layer)
+	layer.blocks = append(layer.blocks, block)
+	s.blockRefs[block.id] = block
+	s.updateRefHeight(layer, block)
+}
+
+func (s *state) updateRefHeight(layer *layerInfo, block *blockInfo) {
+	if layer.verifying.referenceHeight == 0 && layer.lid.After(s.evicted) {
+		layer.verifying.referenceHeight = s.layer(layer.lid.Sub(1)).verifying.referenceHeight
```

Comment:
say layer.lid.Sub(1) is an empty layer (no blocks). so addBlock() is not called for lid.Sub(1)
in that case, should the code keep going back in layers until it finds a non-zero layer reference height?

Reply:
looks like it is not robust indeed

Reply:
added function to find weight from the last non-empty layer (last commit)